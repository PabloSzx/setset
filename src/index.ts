export * from './manager'
export * from './settings'
export * from './static'
