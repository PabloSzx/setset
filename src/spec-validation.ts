import * as Lo from 'lodash'
import { inspect } from 'util'
import {
  appendPath,
  isNamespaceSpecifier,
  isRecordSpecifier,
  renderPath,
  Specifier,
  SpecifierNamespace,
  TraversalInfo,
} from './settings'

/**
 * Validate the spec for basic invariants.
 * todo mapData mapEntryData fixup etc.
 */
export function validateSpecifier(specifier: Specifier, info: TraversalInfo) {
  if (isNamespaceSpecifier(specifier)) {
    Lo.forOwn(specifier.fields, (fieldSpecifier: Specifier, fieldName: string) => {
      validateSpecifier(fieldSpecifier, appendPath(info, fieldName))
    })
    return
  }

  if (isRecordSpecifier(specifier)) {
    validateSpecifier(specifier.entry as SpecifierNamespace, info)
    return
  }

  if (specifier.mapType !== undefined && typeof specifier.mapType !== 'function') {
    throw new Error(
      `Type mapper for setting "${renderPath(
        info
      )}" was invalid. Type mappers must be functions. Got: ${inspect(specifier.mapType)}`
    )
  }
}
