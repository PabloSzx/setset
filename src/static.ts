/**
 * This module only deals with the type level (no term level)
 * It contains the rich conditional type system powering the settings library.
 */

import { Primitive } from 'type-fest'
import {
  ExcludeUndefined,
  IncludesPlainObjectOrInterface,
  IncludesRecord,
  IsEqual,
  IsRecord,
  KeepOptionalKeys,
  KeepRequiredKeys,
  Lookup,
  OnlyPlainObjectOrInterface,
  PlainObject,
} from './utils'

// todo implement context
type BaseContext = { path: string[] }

export type Spec<Input, Data> = NamespaceSpec<Input, Data, {}>

export type InferDataFromInput<Input> = {
  [K in keyof Input]-?: Input[K] extends Leaf
    ? ExcludeUndefined<Input[K]['type']>
    : IncludesPlainObjectOrInterface<Input[K]> extends true
    ? InferDataFromInput<OnlyPlainObjectOrInterface<Input[K]>>
    : ExcludeUndefined<Input[K]>
}

export type UserInput<Input> = UnwrapSyntheticLeavesDeeply<Input, true>

export type KeysWhereNotPresentOrRequiredInData<Input, Data> = {
  [K in keyof Input]: K extends keyof Data ? (undefined extends Data[K] ? never : K) : K
}[keyof Input]

type KeysWhereNotPresentOrOptionalInData<Input, Data> = {
  [K in keyof Input]: K extends keyof Data ? (undefined extends Data[K] ? K : never) : K
}[keyof Input]

type NO_DATA_MATCH = 'NO_DATA_MATCH'

type Node<Input, Data, Key, LocalContext> = Lookup<Input, Key> extends Leaf
  ? LeafSpec<Lookup<Input, Key>['type'], Lookup<Data, Key>>
  : IncludesRecord<Lookup<Input, Key>> extends true
  ? RecordSpec<Lookup<Input, Key>, Lookup<Data, Key>>
  : IncludesPlainObjectOrInterface<Lookup<Input, Key>> extends true
  ? NamespaceSpec<Lookup<Input, Key>, Lookup<Data, Key, NO_DATA_MATCH>, LocalContext>
  : LeafSpec<Lookup<Input, Key>, Lookup<Data, Key>>

/**
 * Namespace Utils
 */

export type ExcludeShorthand<T> = Exclude<OnlyPlainObjectOrInterface<T>, Leaf>

type NormalizeNamespaceInput<Input, Data> = NormalizeNamespaceInputFieldsOptionalityWithDataFieldsOptionality<
  ExcludeShorthand<Input>,
  Data
>

type NormalizeNamespaceInputFieldsOptionalityWithDataFieldsOptionality<Input, Data> =
  // input optional + data optional = input optional
  // input optional + data required = input required
  // input optional + no data       = input required
  // input required + *             = input required
  // prettier-ignore
  (InputKeysWhereOptionalAndInDataToo<Input,Data> extends never ? {} : Pick<Input, InputKeysWhereOptionalAndInDataToo<Input,Data>>) &
  (Required<Omit<Input, InputKeysWhereOptionalAndInDataToo<Input,Data>>>)

type InputKeysWhereOptionalAndInDataToo<Input, Data> = {
  [K in keyof Input]: K extends keyof Data
    ? undefined extends Input[K]
      ? undefined extends Data[K]
        ? K
        : never
      : never
    : never
}[keyof Input]

export type GetNamespaceDataWhereDivergingFromInput<Input, Data> = Pick<
  ExcludeUndefined<Data>,
  GetNamespaceDataKeysWhereDivergingFromInput<
    ExcludeShorthand<ExcludeUndefined<Input>>,
    ExcludeUndefined<Data>
  >
>

export type GetNamespaceDataKeysWhereDivergingFromInput<Input, Data> = {
  [K in keyof Data]: K extends keyof Input
    ? IsNamespaceInputFieldEqualDataField<Data[K], Input[K]> extends true
      ? never
      : K
    : K
}[keyof Data]

// optionality differences does not count in determinining diverging
type IsNamespaceInputFieldEqualDataField<InputField, DataField> = IsEqual<
  ExcludeUndefined<InputField>,
  ExcludeUndefined<DataField>
>

export type OmitNamespaceKeys<T> = {
  [K in keyof T]: IsRecord<ExcludeUndefined<T[K]>> extends true
    ? never
    : HasNamespace<ExcludeUndefined<T[K]>> extends true
    ? never
    : T[K]
}

type HasNamespace<T> = OnlyNamespace<T> extends never ? false : true

export type IsNamespaceInputEqualData<Input, Data> = IsEqual<
  Required<OmitNamespaceKeys<UnwrapSyntheticLeavesDeeply<ExcludeShorthand<Input>>>>,
  Required<OmitNamespaceKeys<ExcludeUndefined<Data>>>
>

type OnlyNamespace<T> = T extends Function
  ? never
  : T extends RegExp
  ? never
  : T extends Date
  ? never
  : T extends any[]
  ? never
  : T extends Leaf
  ? never
  : T extends Primitive
  ? never
  : IsRecord<T> extends true
  ? never
  : T

type OnlyShorthand<T> = UnwrapSyntheticLeaf<ExcludeUndefined<Exclude<T, OnlyNamespace<T>>>>

//prettier-ignore
export type NamespaceSpec<Input, Data, LocalContext> =
  {
    fields: NamespaceFields<ExcludeShorthand<Input>, Data>
  } &
  // todo jsdoc 1) when namespace has no matching data key then developer is responsible
  // todo to map the data tree over _somehow_. Impossible to know how, so return type is
  // todo any possible data. This logic is arbitrary and not guaranteed to work. You should
  // todo unit test it!
  (
    Data extends NO_DATA_MATCH
    ? {}  // something already diverged higher up in the hierarchy, then nothing to do here
    : IsNamespaceInputEqualData<Input, Data> extends true
    // ? { i: Input, d: Data, eq: IsNamespaceInputEqualData<Input, Data> }
    ? {} 
    : { map(normalizedInput: NormalizeNamespaceInput<Input, ExcludeUndefined<Data>>, context: BaseContext & LocalContext): GetNamespaceDataWhereDivergingFromInput<Input, Data> }
  ) &
  /**
   * If namespace is union with non-pojo type then shorthand required 
   */
  (
    OnlyShorthand<Input> extends never
    ? {  }
    : {
        shorthand: Shorthand<OnlyShorthand<Input>, UnwrapSyntheticLeavesDeeply<ExcludeShorthand<Input>>>
      }
  ) &
  /**
   * If namespace is optional AND 1+ sub inputs are required THEN initializer required 
   *  ... if no data match THEN still required
   *  ... but if data optional THEN initial is forbidden (since we'll initialize namespace (data) to undefined)
   *  ... but if all namespace fields (input) are optional THEN initial is forbidden (b/c we can automate
   *      namespace (data) with namespace (input) field initializers)
   */
  (
    undefined extends Input
      ? {} extends KeepOptionalKeys<Input>
        ? {}
        : undefined extends Data
          ? { initial?(): KeepRequiredKeys<Exclude<Input, undefined>> }
          : { initial():  KeepRequiredKeys<Exclude<Input, undefined>> }
      : {}
  )

/**
 * todo
 */
// prettier-ignore
export type NamespaceFields<Input, Data> =
  (
    KeysWhereNotPresentOrOptionalInData<Input, Data> extends never
      ? {}
      : { [Key in KeysWhereNotPresentOrOptionalInData<Input, Data>]+?: Node<Input,Data,Key, {}> }
  ) &
  (
    KeysWhereNotPresentOrRequiredInData<Input, Data> extends never
      ? {}
      : { [Key in KeysWhereNotPresentOrRequiredInData<Input, Data>]-?: Node<Input,Data,Key, {}> }
  )

/**
 * todo: currently assumes Record<string, object>
 */
// prettier-ignore
// todo how does no data match affect this?
export type RecordSpec<Input, Data> =
  (
    {
      /**
       * Specify the settings input for each entry in the record.
       */
      // todo how does no data match affect this?
      entry: Node<
        OnlyPlainObjectOrInterface<ExcludeUndefined<Input>>,
        Data,
        string,
        {
          /**
           * The name under which this entry shows up in the reocrd.
           */
          key: string
        }
      >
    }
  ) &
    /**
     * if input is optional then initial is required
     * unless all
     */
    (
      undefined extends Input
      ? {
          /**
           * Initialize the record with some entries. By default the record will be an empty object.
           */
          initial?(): ExcludeUndefined<Input>
        }
      : {
          /**
           * Initialize the record with some entries. Although you require users to input a record, your initializer will still be run too, if provided.
           */
          initial?(): ExcludeUndefined<Input>
        }
    )

// [1]
// If the field can be undefined it means that initial is not required.
// In most cases it probably means initial won't be supplied. However
// there are may be some odd cases where iniital is present but can
// return undefined.
// prettier-ignore
export type LeafSpec<Input, Data> =
  {
    validate?: Validate<ExcludeUndefined<Input>>
    /**
     * Specify a fixup for this setting.
     *
     * A "fixup" corrects minor problems in a
     * given setting. It also provides a human readable message about what was
     * done and why.
     *
     * Return null if no fixup was needed. Return a fixup object
     * otherwise. The new value should be returned along with a list of one or
     * more messages, one for each thing that was fixed.
     */
    fixup?: Fixup<ExcludeUndefined<Input>>
  } &
  /**
   * if input is optional then initial is required
   * if input is optional and no matching data key then initial is required
   * if input is optional and matching data key optional then initial is optional
   */
  (
    undefined extends Input
      ? undefined extends Data
        ? { initial?(): Input }
        : { initial(): ExcludeUndefined<Input> }
      : {}
  )

/**
 * Isolated Types
 */

export type Validate<Input = Primitive> = (value: Input) => null | { reasons: string[] }

//todo support no messages
export type Fixup<Input = Primitive> = (input: Input) => null | { value: Input; messages?: string[] }

export type MapType<Input = Primitive, Return = Primitive> = (input: Input) => Return

export type Shorthand<Input = Primitive, Return = PlainObject> = (input: Input) => Return

/**
 * Synthetic Leaves
 */

export type Leaf<T = any> = { __settingKind: 'Leaf'; type: T }

export type AlreadyNativeTypeError = 'Error: You wrapped Leaf<> around this field type but Setset already considers it a leaf.'

export type NativeLeaf = Primitive | Date | RegExp

export type UnwrapSyntheticLeavesDeeply<T, CheckUselessWraps extends boolean = false> = {
  [K in keyof T]: T[K] extends Leaf
    ? UnwrapSyntheticLeaf<T[K], CheckUselessWraps>
    : HasNamespace<T[K]> extends true
    ? UnwrapSyntheticLeavesDeeply<T[K]>
    : T[K]
}

export type UnwrapSyntheticLeaf<T, CheckUselessWraps extends boolean = false> = T extends Leaf
  ? CheckUselessWraps extends true
    ? T['type'] extends NativeLeaf
      ? AlreadyNativeTypeError
      : T['type']
    : T['type']
  : T
