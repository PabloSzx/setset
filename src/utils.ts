import { Primitive } from 'type-fest'

/**
 * Check if curerntly in production mode defined as
 * NODE_ENV environment variable equaling "production".
 */
export function isProduction(): boolean {
  return process.env.NODE_ENV === 'production'
}

/**
 * Check if curerntly in development mode defined as
 * NODE_ENV environment variable not equaling "production".
 */
export function isDevelopment(): boolean {
  return process.env.NODE_ENV !== 'production'
}

export type Lookup<Object, Key, Fallback = never> = Key extends keyof Object ? Object[Key] : Fallback

export type IsSameKeys<A, B> = keyof A extends keyof B ? (keyof B extends keyof A ? true : false) : false

export type AnyObject = { [k: string]: any }

export declare type ExcludeUndefined<A> = A extends undefined ? never : A

export type IncludesVoid<T> = (T extends void ? true : never) extends never ? false : true

export type IncludesUndefined<T> = (T extends undefined ? true : never) extends never ? false : true

export type IsRecord<T> = string extends keyof T ? true : false

export type OnlyIndexedType<T> = string extends keyof T ? T : never

export type IncludesRecord<T> = OnlyIndexedType<Exclude<T, Primitive | void>> extends never ? false : true

export type MaybePromise<T = void> = T | Promise<T>

export type CallbackRegistrer<F> = (f: F) => void

export type SideEffector = () => MaybePromise

export type OnlyPlainObjectOrInterface<T> = T extends Function
  ? never
  : T extends RegExp
  ? never
  : T extends Date
  ? never
  : T extends any[]
  ? never
  : T extends Primitive
  ? never
  : T

export type IncludesPlainObjectOrInterface<T> = OnlyPlainObjectOrInterface<T> extends never ? false : true

export type ExcludePlainObjectOrInterface<T> = Exclude<T, OnlyPlainObjectOrInterface<T>>

/**
 * Represents a POJO. Prevents from allowing arrays and functions
 *
 * @remarks
 *
 * TypeScript interfaces will not be considered sub-types. For that, use AnyObject
 */
export type PlainObject = {
  [x: string]: Primitive | object
}

export type isPlainObject<T> = T extends PlainObject ? true : false

export type NotPlainObject = Primitive | any[] | Function

export type IncludesPlainObject<T> = Only<T, PlainObject> extends never ? false : true

export type Only<T, U> = Exclude<T, Exclude<T, U>>

export type KeepOptionalKeys<t> = {
  [k in keyof t]: undefined extends t[k] ? t[k] : never
}

export type GetRequiredKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T]

export type KeepRequiredKeys<t> = {
  [k in ExcludeUndefined<GetRequiredKeys<t>>]-?: undefined extends t[k] ? never : t[k]
}

export type UnknownFallback<T, U> = unknown extends T ? U : T

export type Includes<T, U> = (T extends U ? T : never) extends never ? false : true

export type ExcludePrimitive<T> = Exclude<T, Primitive>

export function assertTypeEquals<T>(expression: T): void {}

export function mergeShallow(o1: any, o2: any) {
  for (const [k, v] of Object.entries(o2)) {
    if (v !== undefined) {
      o1[k] = v
    }
  }
  return o1
}
