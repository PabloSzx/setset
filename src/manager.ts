import * as Logger from '@nexus/logger'
import * as Lo from 'lodash'
import { DataDefault, MetadataState, Spec } from '.'
import { commit, createInfo, dataFromMetadata, FixupInfo, initialize, resolve } from './settings'
import { validateSpecifier } from './spec-validation'
import { isDevelopment, PlainObject } from './utils'

const log = Logger.log.child('settings')

/**
 * todo
 */
export type Manager<Input extends PlainObject, Data extends PlainObject> = {
  reset(): Manager<Input, Data>
  change(input: Input): Manager<Input, Data>
  original(): Data
  metadata: MetadataState<Data>
  data: Data
}

export type Options = {
  /**
   * Handle fixup events.
   *
   * If your settings spec has no fixups then you can ignore this option.
   *
   * By default, fixups are logged at warning level. If you provide your own
   * function then this default behaviour will be disabled. You can retain it by
   * calling the default function passed as a second argument to your function.
   */
  onFixup?: (info: FixupInfo, originalHandler: (info: FixupInfo) => void) => void
}

/**
 *
 */
export function create<Input extends PlainObject, Data extends PlainObject = DataDefault<Input>>({
  fields,
  ...options
}: {
  fields: Spec<Input, Data>
} & Options): Manager<Input, Data> {
  log.debug('construct')
  const info = createInfo()

  if (isDevelopment()) {
    validateSpecifier({ fields }, info)
  }

  // todo we currently have to clone the given spec deeply because mapEntryData mutations the spec with shadow specifiers
  // and shodow specifiers currently break the second+ initialize run (e.g. during reset)
  // todo we didn't catch this yet with own unit test, but other nexus unit tets caught the issue

  const initial = initialize({ fields: Lo.cloneDeep(fields) }, info)
  const state = {
    data: initial.data as Data,
    original: (undefined as any) as Data, // lazy
    metadata: initial.metadata as any, // Metadata<Data>,
  }

  const api: Manager<Input, Data> = {
    data: state.data,
    metadata: state.metadata,
    change(input) {
      log.debug('change', { input })
      const newData = resolve(options, 'change', { fields }, input, state.data, state.metadata, info)
      commit({ fields: Lo.cloneDeep(fields) }, 'change', newData, state.data, state.metadata)
      return api
    },
    reset() {
      log.debug('reset')
      const initial = initialize({ fields: Lo.cloneDeep(fields) }, info)
      api.data = state.data = initial.data as any
      api.metadata = state.metadata = initial.metadata as any
      return api
    },
    original() {
      log.debug('get original')
      const original = state.original ?? dataFromMetadata(state.metadata, {})
      return original
    },
  }

  return api
}
