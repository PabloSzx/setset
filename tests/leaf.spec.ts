import { log } from '@nexus/logger'
import * as S from '../src'
import { c } from './__helpers'

describe('change', () => {
  it('.data reflects the change', () => {
    const s = S.create<{ a: number }>({ fields: { a: {} } })
    expect(s.change({ a: 1 }).data).toEqual({ a: 1 })
  })
  it('.metadata reflects the change', () => {
    const s = S.create<{ a: number }>({ fields: { a: {} } })
    expect(s.change({ a: 1 }).metadata.fields.a).toMatchObject({ from: 'change', value: 1 })
  })
  it('array data is not merged in .data', () => {
    const s = S.create<{ a: number[] }>({ fields: { a: {} } })
    expect(s.change({ a: [2] }).data).toEqual({ a: [2] })
    expect(s.change({ a: [3] }).data).toEqual({ a: [3] })
  })
  it('array data is not merged in .metadata', () => {
    const s = S.create<{ a: number[] }>({ fields: { a: {} } })
    expect(s.change({ a: [3] }).data).toEqual({ a: [3] })
    expect(s.change({ a: [2] }).metadata.fields.a).toMatchObject({ from: 'change', value: [2] })
  })
})

describe('specifiers', () => {
  describe('encounter static errors', () => {
    it('if missing for a fields in the input type', () => {
      S.create<{ a: number }>({ fields: { a: {} } })
      // @ts-expect-error
      S.create<{ a: number }>({ fields: {} })
    })
  })
  describe('with optional data', () => {
    it('can be omitted', () => {
      const s = S.create<{ a: number }, { a?: number }>({ fields: {} })
      expect(s.data.a).toBeUndefined()
    })
    it('can be omitted: and can still be changed', () => {
      const s = S.create<{ a: number }, { a?: number }>({ fields: {} })
      s.change({ a: 1 })
      expect(s.data).toEqual({ a: 1 })
    })
    it('can still be provided too', () => {
      const s = S.create<{ a: number }, { a?: number }>({ fields: { a: {} } })
      expect(s.data.a).toBeUndefined()
    })
  })
})

describe('initial()', () => {
  describe('throws an error', () => {
    it('when not assigned a function', () => {
      expect(() =>
        // @ts-expect-error
        S.create<{ a?: number }>({ fields: { a: { initial: 1 } } })
      ).toThrowErrorMatchingSnapshot()
    })
    it('gracefully upon an unexpected error', () => {
      expect(() =>
        S.create<{ a?: number }>({
          fields: {
            a: {
              initial() {
                throw new Error('Unexpected error while trying to initialize setting')
              },
            },
          },
        })
      ).toThrowErrorMatchingSnapshot()
    })
  })
  describe('encounter static errors', () => {
    it('when undefined is returned', () => {
      // cannot throw error for this b/c sometimes undefined is acceptable (see other tests)
      // @ts-expect-error
      S.create<{ a?: string }>({ fields: { a: { initial: c(undefined) } } })
    })
    it('when used on an input field that is required', () => {
      // @ts-expect-error
      S.create<{ a: number }>({ fields: { a: { initial: c(1) } } })
    })
    it('when return a type that is not assignable to the input field', () => {
      // @ts-expect-error
      S.create<{ a?: string }>({ fields: { a: { initial: c(1) } } })
      // @ts-expect-error
      S.create<{ a?: string | boolean }>({ fields: { a: { initial: c(1) } } })
    })
  })
  it('are run at create time', () => {
    const s = S.create<{ a?: number }>({ fields: { a: { initial: c(1) } } })
    expect(s.data.a).toEqual(1)
  })
  describe('with optional data field', () => {
    it('can be omitted even if the input is optional, in which case the data initializes to undefined', () => {
      const s = S.create<{ a?: number }, { a?: number }>({ fields: { a: {} } })
      expect(s.data.a).toBeUndefined()
    })
    it('can still be provided, in which case the data initializes to the returned value like usual', () => {
      const s = S.create<{ a?: number }, { a?: number }>({ fields: { a: { initial: c(1) } } })
      expect(s.data.a).toEqual(1)
    })
    it('can still be provided, but can now return undefined', () => {
      const s = S.create<{ a?: number }, { a?: number }>({ fields: { a: { initial: c(undefined) } } })
      expect(s.data.a).toBeUndefined()
    })
  })
  describe('can be for input fields of type', () => {
    it('function', () => {
      const s = S.create<{ a?: (n: number) => number }>({ fields: { a: { initial: c((x) => x + 1) } } })
      expect(s.data.a(1)).toEqual(2)
    })
    it('array', () => {
      const s = S.create<{ a?: 1[] }>({ fields: { a: { initial: c([1]) } } })
      expect(s.data.a).toEqual([1])
    })
    it('null', () => {
      const s = S.create<{ a?: null }>({ fields: { a: { initial: c(null) } } })
      expect(s.data.a).toEqual(null)
    })
    // prettier-ignore
    it('boolean, number, string, literal', () => {
      expect(S.create<{ a?: boolean }>({ fields: { a: { initial: c(true) } } }).data.a).toEqual(true)
      expect(S.create<{ a?: number }>({ fields: { a: { initial: c(1) } } }).data.a).toEqual(1)
      expect(S.create<{ a?: 2 }>({ fields: { a: { initial: c(2) } } }).data.a).toEqual(2)
      expect(S.create<{ a?: string }>({ fields: { a: { initial: c('a') } } }).data.a).toEqual('a')
    })
  })
})

describe('fixup()', () => {
  let logs: jest.Mock
  let logSettingsOriginal: any

  beforeEach(() => {
    logs = jest.fn()
    logSettingsOriginal = {
      output: log.settings.output,
      filter: log.settings.filter.originalInput,
      pretty: log.settings.pretty,
    }
    log.settings({ output: { write: logs }, pretty: false })
  })

  afterEach(() => {
    log.settings(logSettingsOriginal)
  })

  describe('when onFixup handler', () => {
    it('is set then called when fixup fixes something', () => {
      const onFixup = jest.fn()
      const s = S.create<{ path: string }>({
        onFixup,
        fields: {
          path: {
            fixup(value) {
              if (value[0] === '/') return null
              return { messages: ['must have leading slash'], value: `/${value}` }
            },
          },
        },
      })
      expect(s.change({ path: 'foo' }).data).toEqual({ path: '/foo' })
      expect(onFixup.mock.calls).toMatchSnapshot()
    })
    it('fails then it errors gracefully', () => {
      // prettier-ignore
      const onFixup = jest.fn().mockImplementation(() => { throw new Error('Unexpected error!') })
      // prettier-ignore
      const s = S.create<{ path: string }>({ onFixup, fields: { path: { fixup() { return { value: 'foobar', messages: [] } } } } })
      expect(() => s.change({ path: '' })).toThrowError(
        'onFixup callback for "path" failed \nUnexpected error!'
      )
    })
    it('is not called for a fixup that returns null', () => {
      const onFixup = jest.fn()
      const s = S.create<{ path: string }>({ onFixup, fields: { path: { fixup: () => null } } })
      s.change({ path: '' })
      expect(onFixup.mock.calls).toEqual([])
    })
    it('not set then defualt is to log a warning', () => {
      log.settings({ filter: '*@warn' })
      // prettier-ignore
      const s = S.create<{ a: string }>({ fields: { a: { fixup() { return { value: 'fixed', messages: ['...'] } } } } })
      s.change({ a: 'foo' })
      expect(logs.mock.calls).toMatchSnapshot()
    })
    it('not set and no messages provided then default is to do nothing', () => {
      log.settings({ filter: '*@warn' })
      // prettier-ignore
      const s = S.create<{ a: string }>({ fields: { a: { fixup() { return { value: 'fixed' } } } } })
      s.change({ a: 'foo' })
      expect(logs.mock.calls).toMatchSnapshot()
    })
    it('is set then default handler is not run', () => {
      log.settings({ filter: '*@warn' })
      // prettier-ignore
      const s = S.create<{ a: string }>({ onFixup() {}, fields: { a: { fixup() { return { value: 'fixed', messages: ['...'] } } } } })
      s.change({ a: 'foo' })
      expect(logs.mock.calls).toEqual([])
    })
    it('is set it can call the original handler to retain the original base behaviour', () => {
      log.settings({ filter: '*@warn' })
      // prettier-ignore
      const s = S.create<{ a: string }>({ onFixup(info, original) { original(info) }, fields: { a: { fixup() { return { value: 'fixed', messages: ['...'] } } } } })
      s.change({ a: 'foo' })
      expect(logs.mock.calls).toMatchSnapshot()
    })
    it('can receive no messages if a fixup did not provide any', () => {
      const onFixup = jest.fn()
      let info
      // prettier-ignore
      const s = S.create<{ a: string }>({ onFixup, fields: { a: { fixup() { return { value: 'fixed'  } } } } })
      s.change({ a: 'foo' })
      expect(onFixup.mock.calls).toMatchSnapshot()
    })
  })
  it('a namespace with shorthand runs through fixups too', () => {
    const onFixup = jest.fn()
    // prettier-ignore
    const settings = S.create<{ a: number | { a: number } }>({
      onFixup,
      fields: { a: { shorthand: (a) => ({ a }), fields: { a: { fixup: (value) => ({ messages: [`must be 1, was ${value}`], value: 1 }) } } } },
    })
    expect(settings.change({ a: 2 }).data).toEqual({ a: { a: 1 } })
    expect(onFixup.mock.calls).toMatchSnapshot()
  })
  it('if fixup fails it errors gracefully', () => {
    // prettier-ignore
    const s = S.create<{ path: string }>({ fields: { path: { fixup() { throw new Error('Unexpected error!') } } } })
    expect(() => s.change({ path: '' })).toThrowErrorMatchingSnapshot()
  })
  describe('static errors', () => {
    it('fixup return .value prop must match the input type', () => {
      // @ts-expect-error
      // prettier-ignore
      S.create<{ a: 1 }>({ fields: { a: { fixup() { return { value: 2, messages: [] } } } } }).data
    })
  })
  it('initial does not pass through fixup', () => {
    // prettier-ignore
    expect(
      S.create<{ a?: number }>({ fields: { a: { initial: () => 1, fixup() { return { value: 2, messages: [] } } } } }).data
    ).toEqual({ a: 1 })
  })
})

describe('validate()', () => {
  it('if a setting passes validation nothing happens', () => {
    const validate = jest.fn().mockImplementation(() => null)
    const settings = S.create<{ a: string }>({ fields: { a: { validate } } })
    settings.change({ a: 'bar' })
    expect(validate.mock.calls).toEqual([['bar']])
  })
  it('if a setting fails validation then an error is thrown with structured metadata attached', () => {
    const validate = jest.fn().mockImplementation((value) => {
      if (value === 'bar') return { reasons: ['Too long', 'Too simple'] }
    })
    const s = S.create<{ a: string }>({ fields: { a: { validate } } })
    let e
    try {
      s.change({ a: 'bar' })
    } catch (e_) {
      e = e_
    }
    expect(e).toMatchSnapshot()
  })
  it('initial does not pass through validate', () => {
    const validate = jest.fn().mockImplementation((value) => {
      if (value === 'bad') return { reasons: ['foobar'] }
    })
    expect(
      S.create<{ a?: number }>({ fields: { a: { initial: c(1), validate } } }).data
    ).toEqual({ a: 1 })
  })
  it('unexpected validator failures error gracefully', () => {
    const validate = jest.fn().mockImplementation((value) => {
      throw new Error('Unexpected error while trying to validate')
    })
    const s = S.create<{ a: string }>({ fields: { a: { validate } } })
    let e
    try {
      s.change({ a: 'bar' })
    } catch (e_) {
      e = e_
    }
    expect(e).toMatchSnapshot()
    expect(e.info).toMatchSnapshot()
    expect(e.value).toMatchSnapshot()
  })
})
