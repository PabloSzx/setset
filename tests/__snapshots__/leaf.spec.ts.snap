// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixup() a namespace with shorthand runs through fixups too 1`] = `
Array [
  Array [
    Object {
      "after": 1,
      "before": 2,
      "messages": Array [
        "must be 1, was 2",
      ],
      "path": "a.a",
    },
    [Function],
  ],
]
`;

exports[`fixup() if fixup fails it errors gracefully 1`] = `
"Fixup for \\"path\\" failed while running on value '' 
Unexpected error!"
`;

exports[`fixup() when onFixup handler can receive no messages if a fixup did not provide any 1`] = `
Array [
  Array [
    Object {
      "after": "fixed",
      "before": "foo",
      "messages": undefined,
      "path": "a",
    },
    [Function],
  ],
]
`;

exports[`fixup() when onFixup handler is set it can call the original handler to retain the original base behaviour 1`] = `
Array [
  Array [
    "{\\"event\\":\\"One of your setting values was invalid. We were able to automaticlaly fix it up now but please update your code.\\",\\"level\\":4,\\"path\\":[\\"setset\\"],\\"context\\":{\\"before\\":\\"foo\\",\\"after\\":\\"fixed\\",\\"path\\":\\"a\\",\\"messages\\":[\\"...\\"]}}
",
  ],
]
`;

exports[`fixup() when onFixup handler is set then called when fixup fixes something 1`] = `
Array [
  Array [
    Object {
      "after": "/foo",
      "before": "foo",
      "messages": Array [
        "must have leading slash",
      ],
      "path": "path",
    },
    [Function],
  ],
]
`;

exports[`fixup() when onFixup handler not set and no messages provided then default is to do nothing 1`] = `Array []`;

exports[`fixup() when onFixup handler not set then defualt is to log a warning 1`] = `
Array [
  Array [
    "{\\"event\\":\\"One of your setting values was invalid. We were able to automaticlaly fix it up now but please update your code.\\",\\"level\\":4,\\"path\\":[\\"setset\\"],\\"context\\":{\\"before\\":\\"foo\\",\\"after\\":\\"fixed\\",\\"path\\":\\"a\\",\\"messages\\":[\\"...\\"]}}
",
  ],
]
`;

exports[`initial() throws an error gracefully upon an unexpected error 1`] = `
"There was an unexpected error while running the initializer for setting \\"a\\" 
Unexpected error while trying to initialize setting"
`;

exports[`initial() throws an error when not assigned a function 1`] = `"Initializer for setting \\"a\\" was configured with a static value. It must be a function. Got: 1"`;

exports[`validate() if a setting fails validation then an error is thrown with structured metadata attached 1`] = `
[Error: Your setting "a" failed validation with value 'bar':

- Too long
- Too simple]
`;

exports[`validate() unexpected validator failures error gracefully 1`] = `
[Error: Validation for "a" unexpectedly failed while running on value 'bar' 
Unexpected error while trying to validate]
`;

exports[`validate() unexpected validator failures error gracefully 2`] = `
Object {
  "path": Array [
    "__root__",
    "a",
  ],
}
`;

exports[`validate() unexpected validator failures error gracefully 3`] = `"bar"`;
